services/src/main/java/io/druid/cli/InsertSegment.java

```diff
@@ -0,0 +1,137 @@
+
+/**
+ */
+@Command(
+    name = "insert-segment",
+    description = "insert a segment into metadata storage"
+)
+public class InsertSegment extends GuiceRunnable
+{
+  private static final Logger log = new Logger(InsertSegment.class);
+
+  @Option(name = "--workingDir", description = "The directory where segments and their descriptor.json reside", required = true)
+  private String workingDirPath;
+
+  @Option(name = "--updateDescriptor", description = "Update descriptor.json if the location of the segment changes", required = false)
+  private boolean updateDescriptor = true;
+
+  private ObjectMapper mapper;
+  private IndexerMetadataStorageCoordinator indexerMetadataStorageCoordinator;
+
+  public InsertSegment()
+  {
+    super(log);
+  }
+
+  @Override
+  protected List<? extends Module> getModules()
+  {
+    return ImmutableList.<Module>of(
+        new Module()
+        {
+          @Override
+          public void configure(Binder binder)
+          {
+            JsonConfigProvider.bindInstance(
+                binder, Key.get(DruidNode.class, Self.class), new DruidNode("tools", "localhost", -1)
+            );
+          }
+        }
+    );
+  }
+
+  @Override
+  public void run()
+  {
+    final Injector injector = makeInjector();
+    mapper = injector.getInstance(Key.get(ObjectMapper.class, Json.class));
+    indexerMetadataStorageCoordinator = injector.getInstance(IndexerMetadataStorageCoordinator.class);
+    final DataSegmentFinder dataSegmentFinder = injector.getInstance(DataSegmentFinder.class);
+
+    log.info("Start seraching segments under [%s]", workingDirPath);
+
+    Set<DataSegment> segments = null;
+    try {
+      segments = dataSegmentFinder.findSegments(workingDirPath, updateDescriptor);
+    }
+    catch (SegmentLoadingException e) {
+      Throwables.propagate(e);
+    }
+
+    log.info(
+        "Done searching segments under [%s], [%d] segments were found",
+        workingDirPath,
+        segments.size()
+    );
+
+    try {
+      insertSegments(segments);
+    }
+    catch (IOException e) {
+      Throwables.propagate(e);
+    }
+
+    log.info("Done processing [%d] segments", segments.size());
+  }
+
+  private void insertSegments(final Set<DataSegment> segments) throws IOException
+  {
+    final Set<DataSegment> segmentsInserted = indexerMetadataStorageCoordinator.announceHistoricalSegments(segments);
+    for (DataSegment dataSegment : segmentsInserted) {
+      log.info("Sucessfully inserted Segment [%s] into metadata storage", dataSegment.getIdentifier());
+    }
+    final Set<DataSegment> segmentsAlreadyExist = Sets.difference(segments, segmentsInserted);
+    if (!segmentsAlreadyExist.isEmpty()) {
+      for (DataSegment dataSegment : segmentsAlreadyExist) {
+        log.info("Segment [%s] already exists in metadata storage, updating the payload", dataSegment.getIdentifier());
+      }
+      indexerMetadataStorageCoordinator.updateSegmentMetadata(segmentsAlreadyExist);
+    }
```

Comment:
since we are adding part of segments in a different transaction, so there is a possibility that some of the segments made it to db and some not. since this is a tool run manually, hopefully, user should just be able to re-run the tool and things will be ok.